### DISCUSSION TITLE
Different accumulator design

### DISCUSSION THREAD

**Discussion by adiabat** [mit-dci/utreexo#249] [Ideas]
I've mentioned this a couple times, related to proof size reduction, using different tree types, etc.  Figured it'd be a good place to write it out.

### History

Long ago, before the COVID19, utreexo began with a very different design.  There were no swaps, no collapses, no remove transforms.  This is the story of...

## Dead-leaf forest

The earlier design didn't have removes the way utreexo does now, and it was more or less the same as the construction in [[8]](https://eprint.iacr.org/2015/718) cited in the paper.  In that paper's appendix A - Limited Dynamism, the authors say that you can delete stuff by replacing a leaf with ⊥, which we'll call a dead leaf rather than upside-down-T.  (called "tombstones" in other papers I think)

If you just replace all the deletions with tombstones, you pretty quickly get 95% dead leaves.  The forest gets up to 30 high, as there's about a billion total txos ever, compared to 70M utxos.  First thing to notice: Huh, this doesn't sound that bad, right?  Max proof length of 30 instead of 26?  No big deal right?  It's actually a little bit worse since things get so sparse that proofs don't overlap as well.

Which leads to the observation: You can do better than just replacing things with dead leaves.  There are whole huge chunks of dead leaves.  Say we define our dead leaf to be `ffffffff...` 256 1-bits in a row.  If you do that, you'll see that you end up with a lot of proofs that have a first element of `ffffffff...` as it's the sibling of the thing you're proving.  You'll also see a bunch of proofs that have `7f0c0e0f...` as the 2nd element of the proof.  Well that makes sense, `7f0c0e0f...` is the parent of two dead leaves, or hash(`ffff...`, `ffff....`)

Similarly with the parent of two `7f0c0e0f...`s and so on, you can precompute the whole partents of dead leaves up to 64 high or whatever, and recognize them.  To make it even simpler, you can redefine your Parent() function to say that the parent of 2 dead leaves returns the same dead leaf.

What happens then is that you "unearth" these dead leaves all the time when you check proofs.  And what you can do with them is repopulate them.  Just as in utreexo today, you perform deletion first, then addition.  But unlike utreexo today, you don't start adding at `numLeaves`, you start adding at the left-most unearthed dead leaves.  Since the more dead leaves there are, the more likely you'll find one in a proof, and the more clustered they are the more you'll large runs of them, you end up with an equilibrium of dead and live leaves.  In the mainnet runs I did, it was a bit above 50% dead leaves, so in theory you're adding just 1 extra hash to each branch.

Well this seems OK, and could probably work.  One of the disadvantages though is the forest gets messy, as new leaves get inserted all over the place, wherever you happen to unearth old dead leaves.  So you don't seem to get as much proof overlap as we have in the swap-based utreexo.  The swap based also seems "cleaner" as there's no weird dead stuff in the forest; every leaf is a UTXO and when it's gone it's gone.  Old leaves tend to 'bubble' to the left where the proofs are longer.  Overall, the current swap-based accumulator design seems better, right?

Well, maybe.  There's another property the dead leaf forest has that could tip the scales and make it better than swap based.  With the dead leaf forest, leaves never *move*.  That's nice in that you don't have to swap them, so deletion is simpler (no `remTrans2()`), but also  **a leaf can commit to it's own position!**

Why would that matter?  In the current leaf, we have this:
```
type LeafData struct {
	BlockHash [32]byte
	Outpoint  wire.OutPoint
	Height    int32
	Coinbase  bool
	Amt       int64
	PkScript  []byte
}
```

That all gets hashed and becomes the leaf hash, which goes into the forest.  In the dead leaves forest, we could have the serialization include the leaf's own position:

```
type LeafData struct {
	BlockHash [32]byte
	Position  uint64
	Outpoint  wire.OutPoint
	Height    int32
	Coinbase  bool
	Amt       int64
	PkScript  []byte
}
```
I think this extra few bytes of data going into the leaf hash lets us cut proof sizes in half.

## Collisions and Pre-Images

Collision attacks are very relevant to utreexo security.  If the attacker can collide a real leaf hash with a fake one, they can spend the fake one and fool everyone.  They could make a "real" utxo which has some reasonable amount of fund, like 0.00001 BTC, and make a "fake" utxo, with an address they contol and a much larger amount of BTC, like 100,000.  They then vary both the real and fake ones and try to find a case where they both hash to the same thing.  They broadcast a transaction creating the "real" utxo, but then spend the "fake" one.  Since both UTXOs have the same leaf hash, the proofs are the same, so utreexo nodes will see a valid proof for the "real" utxo as a valid proof for the "fake" one, and the attacker can get a bunch of money.

Collision attacks take around 2^(n/2) work for n-bit hashes.  So colliding sha256 should take around 2^128 operations.  128 bits is the (unofficial?) security parameter for bitcoin.  So we're good here.

Collision attacks are much easier than pre-image, or 2nd preimage attacks.  pre-image attacks should take 2^n work, so a pre-image attack for sha256 should take 2^256 work, which is way overkill since there are a bunch of things in bitcoin (like the signatures) that only resist 2^128.  Also in general, when hash functions break, the collision resistance tends to break first, so if you have a system or algorithm that only relies on the pre-image resistance of a hash function, that's seen as a stronger system than one which relies on the collision resistance of the hash function.

While utreexo and merkle tress in general use collision resistance, there are some ways around this with utreexo.  Even in the current LeafData, we have that BlockHash field.  That makes collision attacks much more difficult; the attacker can use an old, known BlockHash for the "fake" utxo, but for the new one they're stuck as they don't know what the next BlockHash will be.  So they have some options:

- Make lots of "fake" utxos and use only those to collide.  This is no longer a collision attack, and is 2nd preimage, which takes 2^n, so that's no good.

- Create both "fake" and "real" utxos to collide in the traditional way (cycle finding, etc).  This is extremely expensive, as the collision candidates on the "real" side cost 2^74 hash operations each.  Instead of 1.  Also each successful "real" candidate that you don't broadcast is a whole block reward the attacker left on the table.  So they should broadcast them!

- Seed a bunch of "real" utxos, wait for them to confirm, then try to collide with them.  Each "real" utxo created helps!  256 of them shaves 8-bits of work off your collision effort.

- Really you can just collide with *any* leaf in the forest.  So maybe spam the utxo set a bit, but you already have 80M candidates, which helps your collision a bunch.

We can maybe reduce hash outputs down from 256 bits since normal collision attacks are no longer possible.  However we can't go down to 16 bytes; there are still *some* attacks possible that are in between the standard 2nd pre-image and collision attacks in difficulty.  Worse, this collision attack gets easier the more utxos there are.

## Dead-leaf forests and collision attacks

This is why having a leaf commit to it's own position is great.  If leaves don't move, and every leaf has it's own position inside it, then the attacker can't collide against any leaf in the forest.  They need to pick some number to put into their "fake" utxo, and once they do, they can only collide with the leaf hash at that position if they want their "fake" utxo proof to be accepted.  If they put in position = 5 for their "fake" utxo, and do manage to find a hash collision, but they collide with the leaf at position 23, node will not accept the proof, as the position in the leaf data and the position in the proof don't match up.

I guess you can't quite make the argument that you're right at 2^n work for a fake proof, as the attacker still does have the option to mine valid headers and drop them.  I should figure out how much that really helps; it doesn't seem like it helps all that much.  The attacker could create 2 blocks at the same height with any difference in the resulting leaf set.  Then when they grind though "fake" utxos they have 2 choices, and they broadcast the one that they collide with, dropping the effort by 1 bit.  But it was 2^75 work to go from 2^128 to 2^127.  Also it's not cumulative; the attack has to succeed in under 10 minutes.  TODO: figure out actual equation for this.

That's the other argument though: that an attacker that can pull off this attack is already so powerful that they can just destroy bitcoin anyway, so who cares about faking a proof.  If an attacker has a million times more sha256 power than the bitcoin network, is that an attacker worth worrying about?  They can already re-org to genesis and rewrite the entire chain faster than they can make a fake utreexo proof.  So maybe that's a better metric than an arbitrary 2^128; we say the proofs can defend against a 99.9999% attack. (much more powerful than a 51% attack).

Anyway.  Stuff to try out!  I still have some code for the old dead-leaf forest we could try out at some point.

**Comment by kcalvinalvin** [comment]
Just a random thought with swap based trees. Do you think it's maybe worth it to move subtrees around (like how cowforest is addressing them now) vs moving individual nodes around.

**Reply by adiabat** [reply]
Subtrees do move around.  Pollard swaps whole subtrees by switching two pointers; cowforest uses subtrees, and even forest on disk does one run of nodes per row.  So it already feels subtree-y, but maybe there are other on disk representations that lend themselves to less disk I/O.

The dead-leaves design here has a lot less disk writing, as there are no swaps.  When dead-leaves are inserted, everything up from there needs to be re-hashed and re-written though.  But you won't get very large subtree swaps as we sometimes see in the current swap based forest.  

There are probably... (er, almost definitely!) ways to avoid large subtree swaps with the current design, as there is a lot of freedom in how we swap things.  ExtractTwins, for example, is totally optional.  Also right now all the deletions and swaps collapse the forest to have no gaps, and then the new leaves are added at the bottom right.  It could just as easily put the new leaves in the gaps, and then collapse, or any mix of the two.  So for example, if it detected that a large subtree swap would happen, it could instead throw in some leaves from the list of additions.  Lots of variants possible.

**Comment by adiabat** [comment]
# Swapless deletion

Anyone who's worked on this codebase knows that swaps are annoying.  They're confusing, things move around in weird ways, sometimes they have to be stashed / collapsed, all sorts of weird things like that.

Recently on eprint by Bolton Bailey and Suryanarayana Sankagiri, https://eprint.iacr.org/2021/340 doesn't go into detail about how they do deletions, but it seems like they just trim the deleted leaves and change the "prefix" number associated with each node.  I don't think you need the prefix numbering, but the idea of moving leaves up seems to work.

So swapless utreexo deletion is perhaps a hybrid of the current utreexo design and the Bailey / Sankagiri design.  Additions work the same as in utreexo currently.  Deletions, however, don't use swaps anymore.

To delete a leaf, promote its sibling to its parent.  So 
DeletedNode.Parent = DeletedNode.Sibling
Simple enough.  This still works with csns as nodes which are proved will have parents available to be reassigned to the sibling's hash.

This would then create "hollowed out" trees on the left as leaves are deleted and work their way up to towards the root, and new leaves are added on the right.  One question is what to do when leaves work their way up to the root, or if a root is deleted.  I'm leaning towards don't do anything, and leave a dead root, but there's other ways to do it.  To illustrate the issue (as well as the new simpler deletion mechanism):
```
12              
|-------\     
08      09      10      
|---\   |---\   |---\ 
00  01  02  03  04  05  06 
```
Say we've got 7 leaves as above.  Then we delete 01, and 00 gets promoted to where 08 was:
```
12              
|-------\     
00      09      10      
        |---\   |---\ 
        02  03  04  05  06 
```
Next we delete 02, and 03 gets promoted:
```
12              
|-------\     
00      03      10      
                |---\ 
                04  05  06 
```
Already we can see that we could "drop" the 12 root down one row, and make it sibling's with 10.  This is possible without moving anything left or right, but we do need to know that 00 and 03 are leaves in order to be able to recognize that 12 can drop.  If we just deleted 02, we know 03 is a leaf since we just promoted it, and we know what 00 is, but we don't know that 00 is a leaf unless we keep additional data beyond the hash.  Either a leaf bit (1 bit) or a max depth beneath node, which seems hard to keep track of.

Lets go another step and delete 00:
```
03              
     
                10      
                |---\ 
                04  05  06 
```
From deleting 00, we know that 03 has no sibling as its been promoted to root position.  We know the value of 03, but without a "leaf bit" we don't know if it's a leaf or not.  If its not a leaf, though, we know it can't have more than 2 leaves under it, since it just moved up.  So it could be paired with 10 instead of leaving it floating here.

Then we could even delete 03, and have the whole tree gone, and could move the 10 tree to the left or something.  I'm leaning towards not doing anything and letting it stay empty, as then when leaves are inserted they can only move up.  Also the insertion position of a txo would be the order it shows up in; numleaves never goes down but just becomes txosEver.  In one way that's bad; there will be more roots as the number of roots will be log(txos) not log(utxos).  But maybe that's not a big deal, storing a few more roots isn't hard.  But it allows for leaves to commit to their own position, or really position prefix, which never changes.  If 03 is in the hash, if it's at the bottom we pick root 0, and go right, right.  If we get to the end before going through all the bits of the leaf position, that's OK, it just went up.  This looks like it provides defense against collision attacks mentioned here https://github.com/mit-dci/utreexo/discussions/249. 

The biggest problem I see is that the way we store forest on disk or in ram doesn't fit well with this swapless tree.  There will be large hollowed out areas that are fine for proofs and pollards, but would be lots of 0s in the forest on disk file.  Maybe cowforest can be better adapted to deal with this.  Or maybe we have to do a pollard-on-disk, which really we should have anyway.  It might not be a big deal to let the forest stay on disk and get bigger / empty for a bridge node.  The forest would get to like 60GB or so (1 billion txos, 32 bytes each, *2 for the tree) but probably doesn't have much i/o.  There are no swaps moving things around, so each deletion only changes ~96 bytes (or only changes 32 bytes if you leave the dead hashes below instead of zeroing them out)  

	

## ---

More thoughts:

No, there's still a lot of disk i/o.  Say we're here:
```
12              
|-------\     
00      09      10      
        |---\   |---\ 
        02  03  04  05  06 
```
And then 00 gets deleted.  09 moves to 12, and then 02, 03 need to move to 00, 09.  Higher up deletions can cause large movements upwards as every node in a subtree needs to move up.  It's not as bad as high swaps in that it's only half as large, but still could cause a good deal of read/write access to the forest.


Another issue - while putting the leaf position into the leaf preimage helps against collision attacks, it's not perfect.  Ideally the attacker would have to pick a specific leaf to attack, and collide with that, restricting them to a single hash to collide with, which is no longer a collision attack at all, but a 2nd preimage attack.  This swapless transform doesn't quite get there, as while the attacker does need to pick a leaf, they can collide with any hash above that leaf to successfully forge a proof.  Eg:

```
12              
|-------\     
08      09        
|---\   |---\   
00  01  02  03  
```
The attacker chooses position 02.  The start hashing with their evil utxo, and they win if they get a matching hash to 02, 09, or 12.  So their attack is 3X more effective.  The attacker will target leafs in the highest tree, and in bitcoin's case where you can get a few billion total txos, you have a height of 32 or so.  So it's not quite 2nd preimage resistance, but 5 bits worse than that. (Maybe -- there could be more math involved.)

There are ways to stop this branch collision attack -- for example, keep another bit along with the hash indicating whether the node is a leaf or not.  Then the attacker can't use a collision with a non-leaf node.  They can still collide upper hashes with each other, though, and you're using a bit; if you just made the hash 1 bit longer that seems just as effective (more I think)  So it doesn't seem that defending against targets in the branch are worth it.  Another question is will there actually be any full depth leaves in the big tree on the left?  In practice there might not be, but we can't rely on that for security.

**Reply by Shymaa-Arafat** [reply]
Well, that's quite a lot to read thoroughly; so let me just tackle a few ideas at glance after "k" told me in #257 about here #249, maybe u can fastly get what is fruitful:

1-The Bolton Bailey video contains the detailed deleting steps not mentioned in the eprint, 
https://youtu.be/HEKtDILPeaI
and admit the long disk accesses u mention here too(says theirs took several days to compile)
![Screenshot_2021-04-22-16-21-26-10](https://user-images.githubusercontent.com/83260375/116293393-1f2ee780-a797-11eb-9d52-edeb7410c1ba.jpg)
![Screenshot_2021-04-22-16-21-32-87](https://user-images.githubusercontent.com/83260375/116293639-674e0a00-a797-11eb-8ff9-9a9f717551bc.jpg)
As u say just trimming (the idea adopted from Ethereum, and as u say too Ethereum dropped or going to drop  recursive Length prefixes RLPs too)
.
2- Since the beginning, noticing that Bitcoin uses trees & Ethereum uses Tries, I had the idea that maybe VEB Trees (Van Emde Boas) may be the halfway in the middle since u could think of them as halving the bit range to smaller problems.
.
3-Going thru ur trees here reminded me of Hu-Tucker Optimal Alphabetic Trees (OAT)
The thought of keeping the order of nodes (22 yrs ago I made O(n), O(UTXOs), deletion/ insertion keeping the optimality cond. and using a simple array of struct instead of ADTs)
-1st could this be better than swapping?yes internal nodes change but leaves in the same order & any 2nodes never combine if there is leaf bet them.
-2nd, is it possible to put some weight to each TXO represents its expected duration before being deleted?

**Reply by BoltonBailey** [reply]
Hey, let me give another version of this Hybrid idea.

Instead of moving the sibling of a deleted leaf to it's parent, you could just replace the deleted leaf with a garbage value.

To use your example
```
12              
|-------\     
08      09      10      
|---\   |---\   |---\ 
00  01  02  03  04  05  06 
```
deleting 00 and 01, we get
```
12              
|-------\     
xx      09      10      
|---\   |---\   |---\ 
xx  xx  02  03  04  05  06 
```
Where xx represents some DEADBEEF value.

We actually did some tests of this - it's a bit less space efficient so we did not include them in the paper. But it seems to be much more disk-efficient, so maybe it's worth considering.

**Reply by adiabat** [reply]
Hi @BoltonBailey , thanks for commenting!  (I think you might have messaged in the IRC channel but I can't respond there as your user logged off)

I think I see what you mean here in that this would reduce disk i/o as nothing needs to "move", it can just overwrite things with the "xx" tombstone value.

One downside is that if we need to include those tombstone values in the proof, the proof gets larger as the branch is longer.  
But including the "xx" value in the proof is pointless as it's a known value, so we can substitute it with some very short value, or make it implicit based on the proof length.

But then there's still the issue that there's more hashing going on, as in this example 12 is still the hash of (xx, 09).  So even if the proof is smaller the hashes to compute would be the same.

But then that hashing *also* is pointless, as we can just define a parent function `parent(xx, val) = val` and `parent(val, xx) = val`

So then really the only different is compaction of the tree structure on disk; from the proof verifier's perspective there would be no difference.  So even though on the prover's side, on disk, it looks like 

```
12              
|-------\     
xx      09     
|---\   |---\    
xx  xx  02  03   
```
once it's sent to the verifier it looks like
```
12              
|-------\     
02      03
```
Then for the prover side it's just a tradeoff between compaction and disk i/o.  That does seem nice & simple to work with on the prover side.

Not sure that's exactly what you were getting at, but this does seem faster on the proving side.

**Reply by adiabat** [reply]
One slight complication:

say we start with a 8 leaf tree:
```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```
and delete 00, 01
```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
xx  xx  02  03  04  05  06  07
```
We can't just leave that as is, because a proof for 03 would be 02, 08, 13.  If we leave 08 the proof will be wrong.  So we delete 08 as well:
```
14
|---------------\
12              13
|-------\       |-------\
xx      09      10      11
|---\   |---\   |---\   |---\
xx  xx  02  03  04  05  06  07
```
we need to re-hash 12 and 14, (in case we need to prove 05, which is 04, 11, 12) and since 12 is the parent of xx, 09, 12 will just equal 09.  Also, 09 will never be read anymore, since it's no longer part of any proof.  We can still do this:

```
14
|---------------\
09              13
|-------\       |-------\
xx      09      10      11
|---\   |---\   |---\   |---\
xx  xx  02  03  04  05  06  07
```
and it works.  The proof for 03 is 02, xx, 13 and we skip the xx.  

So we don't have to "move" things up but do have to "copy" them up.  The 02 03 can stay at the bottom though.  With a very sparse tree where the leaves stay at the bottom of a 30+ tall tree, but there are only 3 or 4 real leaves left, there could be a lot of disk seek & read when creating a proof since you still have to read the dummy xx values off the disk to skip them.

So I'm not sure which results in less disk i/o overall.  This way might have more reads but fewer writes (since 02, 03 can stay in place).  Might have to try both ways to see which is faster, but that might not be too hard to do as from the verifier side it should look the same.

**Reply by BoltonBailey** [reply]
> including the "xx" value in the proof is pointless as it's a known value, so we can substitute it with some very short value, or make it implicit based on the proof length.

This is a good optimization I didn't think of.

> But then that hashing also is pointless, as we can just define a parent function parent(xx, val) = val and parent(val, xx) = val

This is an interesting idea. I am a bit nervous about the soundness of this though. Are there weird attacks where a malicious bridge node can pretend the branch is something different from what it is?

For example, take your tree after the removal of 00 and 01

```
14
|---------------\
09              13
|-------\       |-------\
xx      09      10      11
|---\   |---\   |---\   |---\
xx  xx  02  03  04  05  06  07
```

Let's say 02 is now spent and its the only thing spent in the block. The proof for this would consist of the location and the sibling hashes of the branch so it would be
`(location: 02, hash_list: 13, xx, 03)`

But a malicious node might pretend the tree looked like 
```
14
|---------------\
09              13
|-------\       |-------\
09      xx      10      11
|---\   |---\   |---\   |---\
02  03  xx  xx  04  05  06  07
```

And send the proof

`(location: 00, hash_list: xx, 08, 03)`

Causing a double spend. 

## Edit

After consideration, this is fine if we include the location in the leaf hash.

**Reply by BoltonBailey** [reply]
# Actually

After considering again I think it's not fine. We can't let the root hash be the same value 14 after an update, since then we could package the same UTXOs 00 and 01 into a new block and spend them again right? I think whatever we do we will have to recompute hashes higher up in the tree.

**Reply by adiabat** [reply]
I'm pretty sure it's safe because the way I'm thinking about it, the two models for the tree are the same "logical" tree, just with different "physical" representations.  If the verifying node can't tell which version the prover is using then it should be just as secure.  

In the example:
```
14
|---------------\
09              13
|-------\       |-------\
09      xx      10      11
|---\   |---\   |---\   |---\
02  03  xx  xx  04  05  06  07
```
with a proof for `00` of `xx, 08, 03` -- I'm not sure I understand.  Where is 00? Also 08 doesn't exist in either of the trees.  If the verifyer has the "correct" root 14, the proof of 02:  `03, xx, 13` should work (and the xx doesn't need to be sent).  But hashing `00, 03` together gets... something that's nowhere in the tree.  Maybe we're using different sequences, as `xx, 08, 03` seems either top down or mixed somehow.

Also, yes I definitely agree that any change must result in a changed root hash.  So it's just a matter of what to move around on disk, as the data on disk doesn't have to exactly map to what gets hashes since there can be no_ops present on disk but logically absent from the hash tree.

**Reply by BoltonBailey** [reply]
Sorry about that, I meant to write `(location: 00, hash_list: 13, xx, 03)`. But this attack does not work, since in your design the leafdata includes the location. Indeed, this design choice guarantees that every hash can only exist in a single location in the tree, as long as we don't make the `parent(xx, val) = val` and `parent(val, xx) = val` change.

Here's another optimization you could do. Going back to the original example
```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```
When removing hashes 00 and 01 from the tree, bridge nodes don't actually have to touch locations 00 and 01 on disk. We can simply change out disk to read
```
14
|---------------\
12              13
|-------\       |-------\
xx      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```
And let the xx in position 08 denote that all lower down hashes should be xx as well.

**Reply by BoltonBailey** [reply]
Oh I think I see what you are suggesting now - you are suggesting that in the "Logical tree" leaves do not all have to exist at the same depth. Having two identical values at parent and child locations in the "physical tree" is just how we represent the logical tree without moving things on disk.

I now see what was confusing me about the schematic you drew - i was confused between 14 as a hash and 14 as a location: the hash at location 14 does actually change, because before it was the hash of 12 and 13, whereas now it's the hash of 09 and 13.

Yes, I think this logical representation works perfectly, because the set of values in the accumulator is always the set of valid leafdata values anywhere in the tree.

**Reply by adiabat** [reply]
Yeah sometimes I put a * on numbers for "changed" because the number sometimes means "position", sometimes means "the data that was there moved here".  
I agree that "leave junk on the bottom" is OK as it won't be accessed.  Since logically 14 = hash(09, 13) anyone trying to prove 01 as it no longer contributes to 14.

I think all this works OK without leaves committing to their position though.  In what situations do you think that prevents an attack?

**Reply by BoltonBailey** [reply]
It only prevents an attack in the weird model I was thinking of where the hash tree has all the leaves at the same depth, and the branch to a particular leaf can include multiple copies of the same hash. In the logical model you describe I don't think there are any problems.

**Reply by Shymaa-Arafat** [reply]
I know u r not too enthusiased to my ideas, but I thought I should notify u both (the 2 papers authors) of what I think is a significant piece of information, or observation if u prefer to call it...
-If u review the coinbase TXs all long from Block1 to Block1976331, u'll find that their av lifetime may roughly be above quarter a million, ...
-UTXO of block380 is Unspent till now
-All starting from block 1960000 is Unspent till block 1976331 when I last checked.
.
I know this is a testnet, but coinbase TXs r the same as the real Blockchain.
![IMG_20210522_172734](https://user-images.githubusercontent.com/83260375/119253784-1f15e200-bbb3-11eb-95d1-f053a286ce54.jpg)
![IMG_20210522_172944](https://user-images.githubusercontent.com/83260375/119253792-2806b380-bbb3-11eb-9210-1780dae9c86f.jpg)
![IMG_20210522_173206](https://user-images.githubusercontent.com/83260375/119253797-2d63fe00-bbb3-11eb-82ca-ce0332b14ae2.jpg)
![IMG_20210522_173657](https://user-images.githubusercontent.com/83260375/119253799-3228b200-bbb3-11eb-88f6-7e9873ea0d2c.jpg)
.
-I believe it is not statistically correct to embrace different u may say "lifetime distributions" in one power law curve with av lifetime of 2-3.
-Yes coinbase TXs r one in say 2.5k TXs per block nowadays, but they have different characteristics, and they're sure to exist.
-Maybe if we analyzed the rest of TXs types we'll find more anomalies.
.
-First thought ideas would be to categorize transactions, ie UTXOs resulting from them, according to their characteristics.
-Construct different tree for each class, then join roots
-Or maybe build a weighted tree, with weight here means prob to be accessed next block (ie coinbase UTXOs for example have less weight and hence tend to have longer paths in the tree).
-Probably  when spent be replaced by one of the same kind.
....
Sorry for ur time if u don't consider this important & worth discussing.

**Reply by BoltonBailey** [reply]
Yes, in fact, the UTREEXO paper points out there is a spike in 100-block-lifetime UTXOs as compared to 99 and 101 block UTXOs, potentially because of the coinbase effect. Personally, I would avoid treating UTXOs differently because I would assume the extra effort and code to handle them as a special case would not be worth it. Interesting to think about this idea of categorizing transactions. An initial thought I had was to do some kind of machine learning on the transaction graph to keep similar transactions near each other in the tree, (might be more trouble than it's worth again though).

**Reply by Shymaa-Arafat** [reply]
> Yes, in fact, the UTREEXO paper points out there is a spike in 100-block-lifetime UTXOs as compared to 99 and 101 block UTXOs

-99&101 r the smallest values and probably because they're newer blocks(Unspent yet, ie accurately >99 don't know by how much), maybe by now after 10hrs they would grow more.
Block 300 coinbase is not spent till now ie more than 1,976,031 lifetime. I think I'll go to the mining group in bitcointalk & ask them why & when do they usually spend their UTXOs.
.
> the extra effort and code to handle them as a special case would not be worth it.

-What's so hard in putting them in a different tree if u do have a forest anyway?
.
>  An initial thought I had was to do some kind of machine learning on the transaction graph to keep similar transactions near each other in the tree,

-Since there r limited no of  possible TXs, I'd go for a weighted tree more than  m/c learning. I think of it  as a priority queue in  scheduling, or maybe a cache oblivious design? that's just a hunch I'll have to think about it more.

-I think I'll also take a look at other UTXO model Blockchains like cardano how do they handle it?
»»»Update:
They use AVL+ Trees, and have 4 types of UTXOs (but for different reasons)
https://emurgo.io/blog/understanding-unspent-transaction-outputs-in-cardano

**Reply by adiabat** [reply]
Testnet spending data patters are very different than mainnet, so while it's great to use testnet data to make sure things are correct & don't have bugs, it doesn't tell us as much about caching.  Mostly because testnet coins aren't worth anything so most UTXOs / coins are abandoned after a short period of time.  Also the difficulty reset (bug / feature) causes long stretches of 1-tx blocks, and greatly increases the total height of the chain (to nearly 2 million now vs mainnet's ~700K)

Also, UTXOs that never get spent never need to be proven, so those don't matter for proofs.  The metric that matters would be "average lifespan of UTXOs when they are consumed" not "average lifespan of UTXOs in the UTXO set"  -- which in the case of testnet is very different.

We could make some guesses on how long a UTXO will live when it is first created, but for the IBD process everything is already known, so we don't have to guess.

**Reply by Shymaa-Arafat** [reply]
> UTXOs that never get spent never need to be proven, so those don't matter for proofs. 

But surely they affect the height of the tree, and hence other proofs.

> Testnet spending data patters are very different than mainnet

So did u measure it on the real net?
Pls give a resource, link where I can measure it in real net, especially if u haven't calculate it for each kind of TXs separately.

> We could make some guesses on how long a UTXO will live when it is first created, but for the IBD process everything is already known, so we don't have to guess.

The whole point of what I'm saying is to use old data statistics to predict for future data, never said u have to guess in IBD values?!
Maybe u can use IBD as a quicker test of the testnet UTXOs lifetimes; ie for example blocks 300,380 coinbase TXOs when did they got spent or r they still Unspent like those in the testnet?

**Reply by kcalvinalvin** [reply]
> > UTXOs that never get spent never need to be proven, so those don't matter for proofs.
> 
> But surely they affect the height of the tree, and hence other proofs.

UTXOs that are provably unspendable are not added to the tree. This is the same behavior as Bitcoin Core as provably unspendable UTXOs are not added to the UTXO set. So they don't effect the tree in any shape or form.

> > Testnet spending data patters are very different than mainnet
> 
> So did u measure it on the real net?
> Pls give a resource, link where I can measure it in real net, especially if u haven't calculate it for each kind of TXs separately.

You can a bitcoin node to get the data (https://bitcoincore.org/en/download/). The code that we used to parse the data is at https://github.com/mit-dci/utreexo/tree/master/cmd.
 
> > We could make some guesses on how long a UTXO will live when it is first created, but for the IBD process everything is already known, so we don't have to guess.
> 
> The whole point of what I'm saying is to use old data statistics to predict for future data, never said u have to guess in IBD values?!

It's widely known fact that newer UTXOs get spent faster, older ones stay unspent. Chart of this available section 5.3, figure 2 of the Utreexo paper (https://github.com/mit-dci/utreexo/blob/master/utreexo.pdf).

**Reply by Shymaa-Arafat** [reply]
> UTXOs that are provably unspendable are not added to the tree.

I do know about that and I did not include them(I mean even when I used the testnet data)

> It's widely known fact that newer UTXOs get spent faster, older ones stay unspent.

I know about that too, and I clearly expressed that I'm talking UTXO kind (coinbase, merge-mine,....) not being old or new
-I just re-looked closely on Fig2, it tells order of 10 UTXO have life order of 10⁵, when I found about 20 in a scattered sample of 100.
-However it is from the testnet, I'll re-check the true one.
Thank u very much for the links🙏
.
-By the way, u may like to examine this true figure from
https://t.co/nWhWAMsWLs?amp=1
Where 2m TXOs got spent in 2weeks (1-15 May21), 1.4m of them in the 1st five days 1-5 May 2021

![Screenshot_2021-05-25-07-57-54-60](https://user-images.githubusercontent.com/83260375/119460924-7779ea80-bd3f-11eb-8583-a0c3d6e35855.jpg)

**Reply by Shymaa-Arafat** [reply]
I think u should know that I tried the real data from
https://m.btc.com/
-Yes min(not max or av) lifetime of coinbase UTXOs is 101, but THERE ARE very large values
.
-The block 266668 merge-mine TX (the name from the Transaction Graph Analysis paper) is real, 20 coinbase UTXOs are input to one UTXO, they're from 100,000 to 108,000; meaning they all had a lifetime above 156,668
https://m.btc.com/00000000000000066283fa21e3d99992e9a3eef59aa5f423928d44084719166b
follow the 1000Btc Tx to check birth time of each input
https://m.btc.com/f3e6066078e815bb24db0dfbff814f738943bddaaa76f8beba360cfe2882480a

.
-Block 680,000 from last April 2021 coinbase UTXO got spent after 923 block, 681000&681500 Unspent yet till block 685000+
.
-However, I can't figure out yet the reason of the 2m drop in the no of UTXOs. Block 681500 is time-stamped 2May and most TXs l checked r regular 1:2 ( should increase the counter by 1)
https://m.btc.com/00000000000000000004b3f918b242c413ce24e2a73d012cc85f81538febc6e4

**Reply by Shymaa-Arafat** [reply]
I found out there has been some research on predicting UTXOs by Stanford on 2015, 
"Bitcoin UTXO Lifespan Prediction",
 Robert Konrad& Stephen Pinto
, Dec11, 2015
(6 pages report) although no much add ups on it, results r somehow different
![Screenshot_2021-05-28-14-31-30-44_e2d5b3f32b79de1d45acd1fad96fbb0f](https://user-images.githubusercontent.com/83260375/119986617-5d047300-bfc4-11eb-942c-87c632113a04.jpg)
![Screenshot_2021-05-28-14-32-13-41_e2d5b3f32b79de1d45acd1fad96fbb0f](https://user-images.githubusercontent.com/83260375/119986654-68579e80-bfc4-11eb-9725-fc2e1edc452c.jpg)
![Screenshot_2021-05-28-14-41-20-35_e2d5b3f32b79de1d45acd1fad96fbb0f](https://user-images.githubusercontent.com/83260375/119986683-70afd980-bfc4-11eb-9876-d904cbcdd0e5.jpg)
![Screenshot_2021-05-28-14-44-46-80_e2d5b3f32b79de1d45acd1fad96fbb0f](https://user-images.githubusercontent.com/83260375/119986701-786f7e00-bfc4-11eb-87e8-82afea8106bf.jpg)
Ofcourse a lot happened since 2015 that changed a few things, but as I said I can still find UTXOs that live for more than 1000 blocks ....
-Check the first input at this TX (in block 681807
https://www.blockchain.com/btc/tx/2f9cdeb9f58c9ac2975a5a9e6e439dcabd9b7b67989e2a2dcf82998040278ae6
if click on the 1st UTXO, u'll find that it came from this block 669668
https://www.blockchain.com/btc/tx/3cb30629023073c61793c02add6d07e87841008523dd41c1c0d8a74d4be105f2
Lifetime=12,139
.
I reached for 681807 because I was trying to check the sudden drop in the number of UTXOs, so a one that was created in block 681807 got spent in block 682324 with shorter lifetime yes but still= 324+193=517
.
(I think the sudden drop in no was planned from Feb, someone was repeatedly doing what some paper once called merge-mine TX, ie adding up all his/her Btc in one UTXO
.
u should also view this as an insightful reason or justify for the observed long lifes
![IMG_20210528_200037](https://user-images.githubusercontent.com/83260375/120058525-bb286900-c04b-11eb-8706-b195dcabe1ed.jpg)
from a paper about POB
proof of Burn
https://youtu.be/Q5L8-GJVmZw

**Comment by Shymaa-Arafat** [comment]
I'm kind of mixed up where to discuss this better here #249 or #257, 
First is the "Hybrid design" just an idea or it's already approved????
I take it from here it is still in the brain storming phase, but from "Kcalvinalvin" words in  #257 

> With a swapless design, the locality (newest leaves being on the rightmost side) will improve even more

Do u already have a swapless design rightnow?
